### ISSUE TITLE
Puma::Launcher#prune_bundler should load target path instead of symlink for puma_lib_dir

### ISSUE THREAD

**Issue by Martin288** [puma/puma#1272]
I found there was a periodic restart failed in my mina deployment. The main error could be found in puma error log:

> /xxx/.rvm/rubies/ruby-2.4.0/bin/ruby: No such file or directory -- /xxx/project/releases/38/vendor/bundle/ruby/2.4.0/gems/puma-3.8.1/bin/puma-wild (LoadError)

### Steps to reproduce

1) run `mina deploy` multiple times (depended on your `keep_releases` settings in mina config)

### Expected behavior

Puma should be restarted normally.

### Actual behavior
Puma failed to restart, and found error shown above.

### System configuration

**Ruby version**: 2.4.0
**Rails version**: 5.0.1
**Deployment Tool**: Mina
**puma**: 3.8.1
**mina-puma**: 1.0.1
**mina**: 1.0.6

### My Solution
According to the error message, it complained a missing file under a removed **release** directory, so I tried to debug and print some key variables and eventually found that the root reason for it: Puma::Launcher#prune_bundler load gem's required paths which are symlink.

```ruby
def prune_bundler
  return unless defined?(Bundler)
  puma = Bundler.rubygems.loaded_specs("puma")
  dirs = puma.require_paths.map { |x| File.join(puma.full_gem_path, x) }
  puma_lib_dir = dirs.detect { |x| File.exist? File.join(x, '../bin/puma-wild') }
```
`dirs` will be something like: 
```ruby
["/my_application_path/releases/194/vendor/bundle/ruby/2.4.0/gems/puma-3.8.1/my_application_path/releases/194/vendor/bundle/ruby/2.4.0/extensions/x86_64-linux/2.4.0/puma-3
.8.1",
"/my_application_path/releases/194/vendor/bundle/ruby/2.4.0/gems/puma-3.8.1/lib"]
```
But it'd better to use the target's absolute path directly:

```ruby
dirs = puma.require_paths.map { |x| File.join(Pathname(puma.full_gem_path).realpath, x) }
```
and then we will successfully get `dirs` like:
```
["/my_application_path/shared/vendor/bundle/ruby/2.4.0/gems/puma-3.8.1/my_application_path/releases/194/vendor/bundle/ruby/2.4.0/extensions/x86_64-linux/2.4.0/puma-3.8.1",
 "/my_application_path/shared/vendor/bundle/ruby/2.4.0/gems/puma-3.8.1/lib"]
```

I wonder if someone else meet the same problem? I will submit a PR if you guys think it is necessary.

**Comment by nateberkopec** [comment]
I think this change would be OK. Please open a PR.

**Comment by stereobooster** [comment]
I would also like to see full integration test with symlink deployment and rolling restart. This what I'm struggling right now.

- test that code reloads
- test that gem reloads
- test that env variables reloads
- test that remove of previous release folder will not break server

**Comment by stereobooster** [comment]
Also note: this fix will work only for this kind of setup: when `vendor` folder is symlinked to `shared/vendor`

**Comment by x-yuri** [comment]
I'm experiencing related issue right now. I'm not running cluster, and it fails in [the other][2] place. Since it tries to `exec` `puma` from removed release. Let us follow the stack up:

[2]: https://github.com/puma/puma/blob/v3.10.0/lib/puma/launcher.rb#L244

[lib/puma/launcher.rb: restart_args](https://github.com/puma/puma/blob/v3.10.0/lib/puma/launcher.rb#L205)
[lib/puma/launcher.rb: generate_restart_data](https://github.com/puma/puma/blob/v3.10.0/lib/puma/launcher.rb#L370)

At this point we realize, that `puma` gets `/home/user/site/releases/42/...` path from `bundler` (`$0`).

[lib/bundler/cli/exec.rb: kernel_load](https://github.com/bundler/bundler/blob/v1.16.0/lib/bundler/cli/exec.rb#L68)
[lib/bundler/cli/exec.rb: run](https://github.com/bundler/bundler/blob/v1.16.0/lib/bundler/cli/exec.rb#L28)
[lib/bundler.rb: which](https://github.com/bundler/bundler/blob/v1.16.0/lib/bundler.rb#L375) (`ENV['PATH']` is wrong)
[lib/bundler/shared_helpers.rb: set_path](https://github.com/bundler/bundler/blob/v1.16.0/lib/bundler/shared_helpers.rb#L309)
[lib/bundler.rb: bundle_path](https://github.com/bundler/bundler/blob/v1.16.0/lib/bundler.rb#L80)
[lib/bundler.rb: root](https://github.com/bundler/bundler/blob/v1.16.0/lib/bundler.rb#L218)
[lib/bundler/shared_helpers.rb: root](https://github.com/bundler/bundler/blob/v1.16.0/lib/bundler/shared_helpers.rb#L27)
[lib/bundler/shared_helpers.rb: find_gemfile](https://github.com/bundler/bundler/blob/v1.16.0/lib/bundler/shared_helpers.rb#L245)
[lib/bundler/shared_helpers.rb: find_file](https://github.com/bundler/bundler/blob/v1.16.0/lib/bundler/shared_helpers.rb#L253)
[lib/bundler/shared_helpers.rb: search_up](https://github.com/bundler/bundler/blob/v1.16.0/lib/bundler/shared_helpers.rb#L266)
[lib/bundler/shared_helpers.rb: pwd](https://github.com/bundler/bundler/blob/v1.16.0/lib/bundler/shared_helpers.rb#L71)

That's basically it. And I don't see a good solution to the issue, do you? `bundler` is not obliged to use version of the path with a symlink. On the other hand,`puma` can totally expect that the dir exists while it's running.

As a workaround, we can try [`disable_exec_load`][3] (`bundler`), or [`restart_command`][4] (`puma`). Any other ideas?

@Martin288 Does adding `realpath` really help?

[3]: http://bundler.io/man/bundle-config.1.html#LIST-OF-AVAILABLE-KEYS
[4]: https://github.com/puma/puma/blob/v3.10.0/examples/config.rb#L103

**Comment by Martin288** [comment]
@x-yuri right, will try your way later. I think is PR

**Comment by x-yuri** [comment]
I just checked, `disable_exec_load` doesn't help (not surprisingly). `restart_command` does make it work, but it may lead to duplicating settings. But it as well might not, if you keep everything in config file. I made it work with just `restart_command 'bundle exec puma -d'`.

The other option is to not use restart (only stop + start), but it triggers some errors on stop, although seems to work.

**Comment by tomas** [comment]
@Martin288 Did you manage to solve the issue one way or another? It's 2018 and I'm _still_ running into this... :)

**Comment by x-yuri** [comment]
> Did you manage to solve the issue one way or another? It's 2018 and I'm still running into this... :)

@tomas, can you help us find a solution? You see, when you run `bundle exec puma` (what generally happens during development), `bundler` executes `puma` this way: `~/app/releases/1/vendor/bundle/ruby/2.5.0/bin/puma ARGS`. Some number of deploys into the future, and `~/app/releases/1` gets removed. After which `puma` is supposed to [restart][fork] itself. So it does so as `bundler` did (`~/app/releases/1/... ARGS`). If it were executed like this: `~/app/current/... ARGS`, that would be a different story. But how can `puma` figure out which path you want it to use?

[fork]: https://github.com/puma/puma/blob/v3.11.4/lib/puma/launcher.rb#L245

`bundler` itself builds path to `puma` based on current directory. You could say, You're in `~/app/current` when running `bundle exec puma`. What's the matter? But:

```sh
#!/usr/bin/env bash
set -eu
rm -rf 1 2 3
mkdir 1
ln -s 1 2
ln -s 2 3
cd 3
ruby --version
ruby -e 'p Dir.pwd'
ruby -rpathname -e 'p Pathname.pwd'
```

Output:

```
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]
"/home/yuri/_/1/1"
#<Pathname:/home/yuri/_/1/1>
```

To give you more details, `bundler` first adds `bundle_path/bin` to `PATH`:

[Bundler::CLI::Exec#run -> SharedHelpers.set_bundle_environment](https://github.com/bundler/bundler/blob/v1.16.2/lib/bundler/cli/exec.rb#L25)
[Bundler::SharedHelpers.set_bundle_environment -> set_path](https://github.com/bundler/bundler/blob/v1.16.2/lib/bundler/shared_helpers.rb#L90)
[Bundler::SharedHelpers.set_path -> Bundler.bundle_path](https://github.com/bundler/bundler/blob/v1.16.2/lib/bundler/shared_helpers.rb#L309)
[Bundler.bundle_path -> Bundler.root](https://github.com/bundler/bundler/blob/v1.16.2/lib/bundler.rb#L80)
[Bundler.root -> SharedHelpers.root](https://github.com/bundler/bundler/blob/v1.16.2/lib/bundler.rb#L218)
[Bundler::SharedHelpers.root -> find_gemfile](https://github.com/bundler/bundler/blob/v1.16.2/lib/bundler/shared_helpers.rb#L27)
[Bundler::SharedHelpers.find_gemfile -> find_file](https://github.com/bundler/bundler/blob/v1.16.2/lib/bundler/shared_helpers.rb#L245)
[Bundler::SharedHelpers.find_file -> search_up](https://github.com/bundler/bundler/blob/v1.16.2/lib/bundler/shared_helpers.rb#L253)
[Bundler::SharedHelpers.search_up -> SharedHelpers.pwd](https://github.com/bundler/bundler/blob/v1.16.2/lib/bundler/shared_helpers.rb#L266)
[Bundler::SharedHelpers.pwd -> Pathname.pwd](https://github.com/bundler/bundler/blob/v1.16.2/lib/bundler/shared_helpers.rb#L71)

Then looks for executable in paths from the variable:

[Bundler::CLI::Exec#run -> Bundler.which](https://github.com/bundler/bundler/blob/v1.16.2/lib/bundler/cli/exec.rb#L26)
[Bundler.which -> ENV['PATH']](https://github.com/bundler/bundler/blob/v1.16.2/lib/bundler.rb#L375-L382)

After that, it `load`'s (optimization) the executable:

[Bundler::CLI::Exec#run -> kernel_load](https://github.com/bundler/bundler/blob/v1.16.2/lib/bundler/cli/exec.rb#L28)
[Bundler::CLI::Exec#kernel_load -> file](https://github.com/bundler/bundler/blob/v1.16.2/lib/bundler/cli/exec.rb#L68)

Here's the script that reproduce the issue I'm experiencing:

```sh
#!/usr/bin/env bash
set -eu

RAILS_VERSION=5.2.0
PUMA_VERSION=3.11.4
STRATEGY=restart
# STRATEGY=disable-exec-load
# STRATEGY=stop-start
DIR=puma-restart-issue

quote_ere() {
    local s=$1 delimiter=${2:-}
    local re='\.|\*|\[|\^|\$|\\|\+|\?|\{|\||\('   # .*[^$\+?{|(
    if (( ${#delimiter} )); then
        re=$re'|\'$delimiter
    fi
    r=$(printf "%s\n" "$s" | sed -E 's/('"$re"')/\\\1/g')
    printf "%s" "${r//$'\n'/\n}"
}

puma_pids() {
    local ruby_pids=$(pgrep -u "$(whoami)" 'ruby|bundle' | paste -sd,)
    if [ "$ruby_pids" ]; then
        ps -p "$ruby_pids" -o pid,args \
            | grep puma \
            | awk '{print $1}' \
            | paste -sd,
    fi
}

puma_status() {
    echo puma pids: `puma_pids`
    bundle exec pumactl status &> /dev/null || {
        echo status: not running
        more log/std* | cat
    }
}

puma_start() {
    if [ $STRATEGY = disable-exec-load ]; then
        export BUNDLE_DISABLE_EXEC_LOAD=1
    fi
    bundle exec puma -d -C "$(readlink -f ..)/current/config/puma.rb"
    while ! [ -e tmp/sockets/pumactl.sock ] || [ "$(bundle exec pumactl status)" != 'Puma is started' ]; do
        sleep 1
    done
}

puma_kill() {
    local pidfile=$1
    if ! [ -e "$pidfile" ]; then
        return
    fi
    pid=$(cat -- "$pidfile")
    kill -- "$pid" &> /dev/null || true
    while kill -0 "$pid" &> /dev/null; do sleep 1; done
}

ruby --version
bundle --version
puma_kill "$DIR/tmp/pids/puma.pid"
echo puma pids: `puma_pids`

if [ "${RECREATE_SITE-}" ]; then
	rm -rf "$DIR"
fi
if ! [ -e "$DIR" ]; then
	mkdir "$DIR"
	pushd "$DIR"
	gem install rails -v "$RAILS_VERSION"

	# create site
	rails "_${RAILS_VERSION}_" new ra1
	mkdir -p log bundle tmp/{pids,sockets}
	cd ra1
	rm -r log
	ln -s ../log log
	ln -s ../../bundle vendor/bundle
	ln -s ../../tmp/sockets tmp/sockets
	ln -s ../../tmp/pids tmp/pids
	q_PUMA_VERSION=`quote_ere "$PUMA_VERSION" /`
	sed -Ei 's/gem.*puma.*/gem "puma", "'"$q_PUMA_VERSION"'"/' Gemfile
	bundle install --path vendor/bundle
    sed -Ei '2s/^/puts "$0: #{$0}"\n/' "$(bundle exec which puma)"
	echo 'ROOT = File.join __dir__, ".."
stdout_path = File.join ROOT, "log/stdout"
stderr_path = File.join ROOT, "log/stderr"
stdout_redirect stdout_path, stderr_path, true
state_path File.join ROOT, "tmp/sockets/puma.state"
pidfile File.join ROOT, "tmp/pids/puma.pid"
activate_control_app "unix://" + File.join(ROOT, "tmp/sockets/pumactl.sock")
' > config/puma.rb
	popd
fi &> /dev/null

cd "$DIR"
rm -rf release{1,2} current log/*

# create release 1
cp -r ra1 release1
ln -sfn release1 current

# start puma
echo
echo -- puma: start
cd current
puma_start
puma_status

# create release 2
cd ..
cp -r release{1,2}
rm -rf release1
ln -sfn release2 current

# restart puma
cd current
echo
echo -- puma: restart
case "$STRATEGY" in
restart|disable-exec-load)
    bundle exec pumactl restart
    sleep 2
    puma_status;;
stop-start)
    bundle exec pumactl stop
    puma_start
    more log/std* | cat;;
esac
```

Output:

```
ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]
Bundler version 1.16.2
puma pids: 

-- puma: start
$0: /home/yuri/_/restart-puma/puma-restart-issue/release1/vendor/bundle/ruby/2.5.0/bin/puma
Puma starting in single mode...
* Version 3.11.4 (ruby 2.5.1-p57), codename: Love Song
* Min threads: 0, max threads: 16
* Environment: development
* Daemonizing...
puma pids: 32578

-- puma: restart
Command restart sent success
puma pids: 
status: not running
::::::::::::::
log/stderr
::::::::::::::
=== puma startup: 2018-07-04 10:36:38 +0300 ===
/home/yuri/.rubies/ruby-2.5.1/bin/ruby: No such file or directory -- /home/yuri/_/restart-puma/puma-restart-issue/release1/vendor/bundle/ruby/2.5.0/bin/puma (LoadError)
::::::::::::::
log/stdout
::::::::::::::
=== puma startup: 2018-07-04 10:36:38 +0300 ===
* Listening on tcp://0.0.0.0:9292
* Starting control server on unix:///home/yuri/_/restart-puma/puma-restart-issue/current/config/../tmp/sockets/pumactl.sock
* Restarting...
```

In case, you're experiencing another issue, please amend the script according to what you see.

**Comment by tomas** [comment]
@x-yuri I ended up solving the issue by hardcoding the `restart_command` path on my `puma.rb` config file. This way the restart task doesn't need to rely on Bundler for figuring out the location of the original command (which no longer exists after a few deploys).

```
root = '/path/to/your/app'
environment ENV['RACK_ENV'] || 'production'

rackup root + '/current/config.ru'
pidfile root + '/shared/pids/puma.pid'
bind 'unix://' + root + '/shared/sockets/puma.sock'
restart_command root + '/current/bin/puma'

workers 128
prune_bundler
```

**Comment by x-yuri** [comment]
128 workers? O.o You're one slave-driver, aren't you? :)

Anyways, the workaround is indeed a `restart_command`. But I'd rather suggest `restart_command 'bundle exec puma'` for being more readable. In case `puma` starts from `current` dir, not from `releases/<n>`.

**Comment by tomas** [comment]
Ha! The number was actually 12. I just thought 128 would look cooler. ;)

**Comment by nateberkopec** [comment]
Linking #1105 as it will have to fix this issue as well.

**Comment by nateberkopec** [comment]
Closed by #1105
